### PR TITLE
perf fix for Ipv4Instance creation

### PR DESCRIPTION
**Pull request by brian-pane**
*Description*:
Speed up the creation of an `Ipv4Instance` from a `sockaddr_in`.
I noticed in a CPU profile that this constructor was taking a surprisingly large amount of time.
This change includes a new performance benchmark test.

Before:
```
Benchmark                   Time           CPU Iterations
----------------------------------------------------------
Ipv4InstanceCreate        474 ns        474 ns    1470706
```
After:
```
----------------------------------------------------------
Benchmark                   Time           CPU Iterations
----------------------------------------------------------
Ipv4InstanceCreate         58 ns         58 ns   12214700
```

*Risk Level*: medium

*Testing*: unit tests plus new performance benchmark test

*Docs Changes*: N/A
*Release Notes*: N/A

Signed-off-by: Brian Pane <brianp+github@brianp.net>

### CODE REVIEW

File: source/common/network/address_impl.cc [hunk 1]
```diff
@@ -22,6 +22,35 @@ namespace Address {
...
+static std::string Ipv4ToString(const sockaddr_in& addr) {
+  static constexpr size_t BufferSize = 16; // enough space to hold an IPv4 address in string form
+  char str[BufferSize];
+  // Write backwards from the end of the buffer for simplicity.
+  char* start = str + BufferSize;
+  uint32_t ipv4_addr = ntohl(addr.sin_addr.s_addr);
+  for (unsigned i = 4; i != 0; i--, ipv4_addr >>= 8) {
+    uint32_t octet = ipv4_addr & 0xff;
+    if (octet == 0) {
+      *--start = '0';
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: ASSERT(start > str) above this, mostly for the reader's benefit.

File: source/common/network/address_impl.cc [hunk 2]
```diff
@@ -22,6 +22,35 @@ namespace Address {
...
+    } else {
+      do {
+        *--start = '0' + (octet % 10);
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: ASSERT(start > str);

File: source/common/network/address_impl.cc [hunk 3]
```diff
@@ -22,6 +22,35 @@ namespace Address {
...
+        octet /= 10;
+      } while (octet != 0);
+    }
+    if (i != 1) {
+      *--start = '.';
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: ASSERT(start > str);

File: source/common/network/address_impl.cc [hunk 4]
```diff
@@ -172,10 +201,12 @@ int InstanceBase::socketFromSocketType(SocketType socketType) const {
...
-  char str[INET_ADDRSTRLEN];
-  inet_ntop(AF_INET, &address->sin_addr, str, INET_ADDRSTRLEN);
-  friendly_name_ = fmt::format("{}:{}", str, ntohs(address->sin_port));
-  ip_.friendly_address_ = str;
+  std::string friendly_address = Ipv4ToString(*address);
+  ip_.friendly_address_ = friendly_address;
+  fmt::format_int port(ntohs(address->sin_port));
+  friendly_name_ = friendly_address;
+  friendly_name_.push_back(':');
+  friendly_name_.append(port.data(), port.size());
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: It looks like there might be one or two extra copies/resizes here, IIUC. Can you use Ipv4ToString to directly write ip_.friendly_address_, and then compose friendly_name_ in one shot:
```
friendly_name_ = absl::StrCat(ip_.friendly_address, ":", ntohs(address->sin_port));
```
Here I'm exploiting the fact that absl::StrCat will do a pass over the args to find the right size, and then do one allocation into the target string before it starts copying bytes.

File: source/common/network/address_impl.cc [hunk 5]
```diff
@@ -22,6 +22,35 @@ namespace Address {
...
 namespace {
 
+/**
+ * Convert an IPv4 address to canonical string format.
+ * @note This works similarly to inet_ntop() but is faster.
+ * @param addr address to format.
+ * @return the address in dotted-decimal string format.
+ */
+static std::string Ipv4ToString(const sockaddr_in& addr) {
+  static constexpr size_t BufferSize = 16; // enough space to hold an IPv4 address in string form
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: can you add unit-tests for this? I can imagine we should hit some corner cases of 0s and 255s in every octet.

I realize you'd have to make this non-static.

File: source/common/network/address_impl.cc [hunk 6]
```diff
@@ -22,6 +22,35 @@ namespace Address {
...
+  uint32_t ipv4_addr = ntohl(addr.sin_addr.s_addr);
+  for (unsigned i = 4; i != 0; i--, ipv4_addr >>= 8) {
+    uint32_t octet = ipv4_addr & 0xff;
+    if (octet == 0) {
+      *--start = '0';
+    } else {
+      do {
+        *--start = '0' + (octet % 10);
+        octet /= 10;
+      } while (octet != 0);
```
Comment (jmarantz) [inline source/common/network/address_impl.cc]: Also worth trying for speed: a 256 element static lookup table. I would declare this as:
```
  static const char* OctedLookup[] = {"0", "1", "2", ...};
```
to get the compiler to lay it out in text-space, as opposed to having it constructed at runtime (and also running into static-initialization fiasco issues).

File: source/common/network/address_impl.cc (hunk 5 above)
Comment (mattklein123) [reply source/common/network/address_impl.cc]: +1 on comprehensive UTs. When we override a stdlib function like this I think we need to backfill comprehensive tests that cover all corner cases.

File: source/common/network/address_impl.cc (hunk 4 above)
Comment (brian-pane) [reply source/common/network/address_impl.cc]: Thanks for the suggestion. Eliminating the temporary friendly_address string provided a slight speedup. StrCat yielded a perf regression (perhaps its int-to-string formatter is slower than fmt::format_int?), but I was able to achieve the same goal of preventing a resize by doing reserve+append on the target string.

Comment (jmarantz) [reply source/common/network/address_impl.cc]: That's surprising. Can you leave a comment describing what you found? IMO the StrCat call is easier to read and is correct by inspection; I'd be curious why it's slow.

Also: it's probably worth just double-checking methodology: is your machine otherwise idle at when running microbenchmarks, and are you repeating them a few times? There can be some variance.

File: source/common/network/address_impl.cc (hunk 6 above)
Comment (jmarantz) [reply source/common/network/address_impl.cc]: did you get a chance to try this?

File: test/common/network/address_impl_test.cc [hunk 7]
```diff
@@ -119,6 +120,22 @@ TEST(Ipv4CompatAddressImplSocktTest, SocketBindAndConnect) {
...
+  // Test addresses from various RF 5735 reserved ranges
+  static const char* addresses[] = {"0.0.0.0",        "0.0.0.255",       "0.0.255.255",
+                                    "0.255.255.255",  "192.0.2.0",       "198.151.100.1",
+                                    "198.151.100.10", "198.151.100.100", "10.0.0.1",
+                                    "10.0.20.1",      "10.3.201.1"};
+
+  for (const auto address : addresses) {
+    sockaddr_in addr4;
+    addr4.sin_family = AF_INET;
+    EXPECT_EQ(1, inet_pton(AF_INET, address, &addr4.sin_addr));
```
Comment (jmarantz) [inline test/common/network/address_impl_test.cc]: this probably should be ASSERT_EQ but in practice it's not a big deal here.

File: test/common/network/address_impl_test.cc [hunk 8]
```diff
@@ -119,6 +120,22 @@ TEST(Ipv4CompatAddressImplSocktTest, SocketBindAndConnect) {
   }
 }
 
+TEST(Ipv4InstanceTest, SockaddrToString) {
+  // Test addresses from various RF 5735 reserved ranges
+  static const char* addresses[] = {"0.0.0.0",        "0.0.0.255",       "0.0.255.255",
+                                    "0.255.255.255",  "192.0.2.0",       "198.151.100.1",
+                                    "198.151.100.10", "198.151.100.100", "10.0.0.1",
+                                    "10.0.20.1",      "10.3.201.1"};
```
Comment (jmarantz) [inline test/common/network/address_impl_test.cc]: can you include a 255 in the first word?  Let's make sure we have testcases where each segment is 0, 255, and something else.

File: test/common/network/address_impl_test.cc [hunk 9]
```diff
@@ -119,6 +120,22 @@ TEST(Ipv4CompatAddressImplSocktTest, SocketBindAndConnect) {
...
+
+  for (const auto address : addresses) {
+    sockaddr_in addr4;
+    addr4.sin_family = AF_INET;
+    EXPECT_EQ(1, inet_pton(AF_INET, address, &addr4.sin_addr));
+    addr4.sin_port = 0;
+    EXPECT_EQ(address, Ipv4Instance::sockaddrToString(addr4));
```
Comment (jmarantz) [inline test/common/network/address_impl_test.cc]: EXPECT_STREQ

File: source/common/network/address_impl.cc (hunk 6 above)
Comment (brian-pane) [reply source/common/network/address_impl.cc]: I won't have time to try a lookup-table implementation, but I won't object if anybody else wants to make a follow-up PR for it.

File: test/common/network/address_impl_test.cc (hunk 9 above)
Comment (brian-pane) [reply test/common/network/address_impl_test.cc]: The existing string checks in address_impl_test.cc all use EXPECT_EQ. What's different about this one that needs EXPECT_STREQ?

Comment (jmarantz) [reply test/common/network/address_impl_test.cc]: It's mostly about legibility of output when there is a failure.  Some forms of comparing char*, char[], string_view, string, etc, result in mismatch displays showing arrays of integers rather than strings.

Comment (brian-pane) [reply test/common/network/address_impl_test.cc]: I just checked, and in this case EXPECT_EQ shows both the left and right sides as strings if they don't match.

Comment (jmarantz) [reply test/common/network/address_impl_test.cc]: OK usually I just recommend using EXPECT_STREQ when comparing all strings, as it's presence never hurts, and its absence can. I also thought maybe a char* to char* comparison might be a pointer-comparison if you're not careful, but obviously you'd know that pretty quick.
